ksp_plugin_adapter/ownership_transfer_marshaler.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Runtime.InteropServices;
+
+namespace principia {
+namespace ksp_plugin_adapter {
+
+// An adapter that uses TMarshaler to marshal objects of type T, but
+// additionally takes ownership of the native data passed in the native-to-
+// managed direction.  T may be a struct or a class, but it *must* correspond to
+// an interchange message defined in journal.proto.
+internal class OwnershipTransferMarshaler<T, TMarshaler> : MonoMarshaler
+    where TMarshaler : MonoMarshaler {
+  public static ICustomMarshaler GetInstance(string s) {
+    return instance_;
+  }
+
+  public override void CleanUpNativeDataImplementation(IntPtr native_data) {
+    t_marshaler_instance_.CleanUpNativeDataImplementation(native_data);
+  }
+
+  public override IntPtr MarshalManagedToNativeImplementation(
+      object managed_object) {
+    return t_marshaler_instance_.MarshalManagedToNativeImplementation(
+        managed_object);
+  }
```

Comment:
This should have the same checks as `OwnershipTransferUTF8Marshaler` (though the error message should more accurately say that we do not support transfer of ownership in instead of suggesting something that might not work).

https://github.com/mockingbirdnest/Principia/blob/2d809b32aa1a9c3849a9b131185bd91779d768fc/ksp_plugin_adapter/utf8_marshaler.cs#L32-L39